pkg/codegen/dotnet/gen_program_test.go

```diff
@@ -7,23 +7,6 @@ import (
 	"github.com/pulumi/pulumi/pkg/v3/codegen/testing/test"
 )
 
-func TestGenerateProgram(t *testing.T) {
```

Comment:
Re: Separate makefile/etc. I want to lean in to using `gotestsum` here because it automatically divides packages across a number of jobs based on previous run data. It's much smarter than us attempting to manually divide up jobs using flags.

Reply:
Here's an example of what gotestsum generated, with text wrapped for readability:

```
    test-command: GO_TEST_TAGS="all" PKGS="github.com/pulumi/pulumi/pkg/v3/codegen/dotnet
      github.com/pulumi/pulumi/pkg/v3/codegen/schema
      github.com/pulumi/pulumi/pkg/v3/codegen/testing/codegen/nodejs1of6
      github.com/pulumi/pulumi/pkg/v3/resource/stack
      github.com/pulumi/pulumi/pkg/v3/codegen/testing/codegen/dotnet3of6
      github.com/pulumi/pulumi/pkg/v3/codegen/importer github.com/pulumi/pulumi/pkg/v3/codegen/pcl
      github.com/pulumi/pulumi/pkg/v3/codegen/testing/codegen/python1of6
      github.com/pulumi/pulumi/pkg/v3/codegen/testing/codegen/python2of6
      github.com/pulumi/pulumi/pkg/v3/backend/display github.com/pulumi/pulumi/pkg/v3/operations
      github.com/pulumi/pulumi/pkg/v3/engine github.com/pulumi/pulumi/pkg/v3/resource/deploy/providers
      github.com/pulumi/pulumi/pkg/v3/util/cancel github.com/pulumi/pulumi/pkg/v3/secrets/b64
      github.com/pulumi/pulumi/pkg/v3/resource/provider github.com/pulumi/pulumi/pkg/v3/version
      github.com/pulumi/pulumi/pkg/v3/backend/state github.com/pulumi/pulumi/pkg/v3/util/testutil
      github.com/pulumi/pulumi/pkg/v3/secrets/service github.com/pulumi/pulumi/pkg/v3/util/tracing
      github.com/pulumi/pulumi/pkg/v3/secrets/cloud
      github.com/pulumi/pulumi/pkg/v3/codegen/hcl2/model/format
      github.com/pulumi/pulumi/pkg/v3/authhelpers
      github.com/pulumi/pulumi/pkg/v3/codegen/testing/codegen-generator
      github.com/pulumi/pulumi/pkg/v3/graph github.com/pulumi/pulumi/pkg/v3/resource/deploy/deploytest
      github.com/pulumi/pulumi/pkg/v3/secrets" ./scripts/retry make gotestsum/pkg
```

Reply:
The other question I have is why we have a separate module for this:
```go
func TestGenerateProgram(t *testing.T) {
   os.Chdir("../../../dotnet") // chdir into codegen/dotnet
   codegen.GenerateProgramBatchTest(t, test.ProgramTestBatch(1, 6))
}
```

Instead of just having 
```go
func TestGenerateProgram(t *testing.T) {
  for i := 0; i < BATCHES; i++ {
    t.Run(fmt.Sprintf("%dof%d", I, BATCHES),
      codegen.GenerateProgramBatch(t, test.ProgramTestBatch(i, BATCHES))
  }
}
```

We can still run them independently as `go test --run=TestGenerateProgram/1of6`.

Does this play better with `gotestsum`? Does separate modules help with process isolation?

Reply:
`gotestsum`'s ci-matrix functionality is constrained by the behavior of `go test`, and so it will generate a matrix with either:

* Package granularity, given a module path it will enumerate packages, placing each into buckets based on previous run duration

  Equivalent to `go test pkg1 pkg2 pkg3 ...`
* Test granularity, given a **single package** it will enumerate tests (`go test -list`), placing each into buckets ...

  `go test -run $REGEX givenPkg`

Reply:
Here's an example of strategy 1, pkg tests 1 of 3: https://github.com/pulumi/pulumi/actions/runs/3269225315/jobs/5376535278

From the "Set up job" step, "Inputs" output:

> test-command: GO_TEST_TAGS="all" PKGS="github.com/pulumi/pulumi/pkg/v3/codegen/dotnet github.com/pulumi/pulumi/pkg/v3/codegen/docs" ./scripts/retry make gotestsum/pkg

And strategy 2, pkg/codegen/nodejs tests 1 of 4: https://github.com/pulumi/pulumi/actions/runs/3269225315/jobs/5376535555

> test-command: GO_TEST_TAGS="all" PKGS="github.com/pulumi/pulumi/pkg/v3/codegen/nodejs" OPTS="-run ^(TestEscape|TestGenerateTypeNames|TestGenerateProgramVersionSelection|TestLazyLoadsGeneration)$" ./scripts/retry make gotestsum/pkg/codegen/nodejs